### ISSUE TITLE
State-API - Inconsistent State-Object when making fast succeeding POST requests (up to 90% missing data. RaceCondition?)

### ISSUE THREAD

**Issue by psacher** [LearningLocker/learninglocker#890]
Hi, 
I think I stumbled over a bug in the State-API. 
Can it be that there isn't any kind of Queue when processing the incoming POST requests???

**Version**
develop at {{06fa1a12152f76a8ce7c084c662bb4f28e899a4e}} on 1.11.0

**Steps to reproduce the bug**
1. send multiple State-Objects with different content to the same State-Document (POST State)
2. do this in close succession without waiting for the response of the server (asynchronously)
3. Check out the "TestSuite" that I have build 
[StateMergeTest.zip](https://github.com/LearningLocker/learninglocker/files/704286/StateMergeTest.zip)
Just enter your LearningLocker credentials and try the two different Send Methods.
The synchronous method works fine, the asynchronous, where we are firing the requests VERY fast fail and the State at the LRS isn't consistent and complete.

**Expected behaviour**
State Objects should be merged correctly and reliably like defined in the [xAPI Spec](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Communication.md#json-procedure-with-requirements)

**Actual behaviour**
The State-Document the LRS has build from the multiple POSTs isn't consistent. 20-90% of the POSTs aren't present in the State-Document. 
The Requests are answered with 204 why I would expect that everything is fine and the information is saved, but it isn't present in the state-object. :(

**Additional information**
OS: Win10
Browser: FireFox, Chrome, IE (all @latestVersion)
I'm using the xAPIWrapper from ADL to send the States to my LL.

I suppose the LearningLocker has no queue or pipe when updating/writing the state-documents and runs into a race condition when getting very fast succeeding update/merge requests.
Just try my TestSuite and you should see the error.
It testet it with the ADL LRS and Grassblade and they worked perfectly. Only LearningLocker fails :(

Thanks for looking into this!

**Comment by psacher** [comment]
Here's an output of my test suite when it bitterly fails :/
![statebug](https://cloud.githubusercontent.com/assets/10605556/21931169/3fdab2b8-d99a-11e6-9b37-16401f06049b.JPG)

**Comment by ryansmith94** [comment]
Hi @psacher, thanks for reporting this bug and using our [template](/contributing.md#issue-templates). We will now try to reproduce your issue.

**Comment by davidpesce** [comment]
I can confirm this doesn't happen on 1.15.0

**Comment by ryansmith94** [comment]
Thanks @davidpesce 👍

**Comment by psacher** [comment]
Thanks for testing. I'll ask my admin for updating our server. Then I will try again.
It would be great, if it is fixed in the current version, then I don't have to implement an own queue in our WBT ^^

**Comment by psacher** [comment]
Hi, 
my admin has just updated to the newest v1.15.0.

The bug is still present :(

![image](https://cloud.githubusercontent.com/assets/10605556/22504845/b12efe00-e878-11e6-9d0a-090c781b8c0a.png)

**Comment by davidpesce** [comment]
@psacher I've got mine set up on a 2GB digitalocean droplet (CentOS). I previously had it on a 1GB droplet without issue. That being said, I'm not pushing a ton of statements to it. The hardest I've pushed it was with your script.

What type of hardware do you have it running on now? Is it under load (lots of statements already coming in) when you run your script?

**Comment by psacher** [comment]
Hi, I don't think that it's a hardware problem. 
LearningLocker and the MongoDB runs flawless at our servers. I just have the problem, that fast successive State posts (that should be merged together) arent in the DB at the end.

As said before, I ASSUME that merge 2,3,4 and 5 arent saved, because merge 1 hasn't finished yet (or something like that).

The depressing fact is, that the post result is 204. So I would think, that all is saved. But it is not. :( 
So I would say, LL fails at reliably merging the states together.

Here is a "good run".

I just made about 50 runs. All looked like this and where incomplete (around 5 of 10 items are in the final state): 
![image](https://cloud.githubusercontent.com/assets/10605556/22552073/a8e4e9c8-e957-11e6-9154-41293b6cf3f4.png)

Then I had luck and in ONE run the state was complete!
![image](https://cloud.githubusercontent.com/assets/10605556/22552370/f6f67bda-e958-11e6-92dd-0f109097d56f.png)

I really don't know what to do anymore. It sucks, that you can't reproduce it. Maybe someone else could test it please?

Bye,
Patrick

**Comment by ht2** [comment]
Hey guys,

The offending line of code lies here: https://github.com/LearningLocker/learninglocker/blob/develop/app/models/DocumentAPI.php#L77

As you suspect @psacher, LL is handling multiple requests simultaneously and some are clearly executing before others. Because LL is fetching the document, then doing the merge on the server and then re-saving the document, you get a race condition where one will overwrite another.

Ideally, instead of the fetch, merge, save it would instead use a $set to simply merge the JSON at the database level therefore ensuring that the resulting document has all the respective keys added, however I I don't think this solves the problem entirely as 2 requests might overwrite each other if the same key is used, but that is simply a limitation of the spec and short of queuing requests in order of receipt at the server level I'm not sure what else could be done about that.

So in summary, a `$set` on update would solve the issue, but there will need to be a bit of a rewrite for that to happen. Certainly one for us to look at eventually but I can't promise HT2 will be jumping to fix this today, tomorrow or even next week (mainly as we have paid client commitments to prioritise). I would however, as always, implore that members of the community look to try and develop and solve problems like this. I would allow that in as a PR in a heartbeat if it was seen to fix the problem.

James

**Comment by psacher** [comment]
Ha!!! THX!
Just that was my assumption.

Thanks for localizing it. :)

Sad to hear, that you won't priorize it. But maybe we will build a hotfix for our WBTs. We'll see. 
THX :D

**Comment by ht2** [comment]
Addressed in v2.0.0

**Comment by psacher** [comment]
I have tried out LearningLocker V2 and I'm happy to say, that everything seems to be working fine now.

I will have an eye on this but it looks very promising. 
Thanks very much for handling this.

**Comment by ht2** [comment]
Excellent! The new API uses the mongo $set to ensure updates are pushed on top of other changes, we also utilise etags correctly to avoid issues with concurrency (as defined by the specification). 👍

**Comment by ryansmith94** [comment]
Hey @psacher, that's great! Thanks for trying it out with V2 and getting back to us, it's really appreciated 😄
